### ISSUE TITLE
Improve exception for One-to-many from Keyless Entity Type 

### ISSUE THREAD

**Issue by optiks** [dotnet/efcore#24283]
## Background
Hi team (long time no see!),

We're trying to configure a one-to-many from a Keyless Entity Type -> regular entity, however we're running into a `NullReferenceException` during model creation.

According to [Keyless Entity Types](https://docs.microsoft.com/en-us/ef/core/modeling/keyless-entity-types?tabs=data-annotations), 

> They can only contain reference navigation properties pointing to regular entities.

[Relationships](https://docs.microsoft.com/en-us/ef/core/modeling/relationships) defines a reference navigation property as, 

> A navigation property that holds a reference to a single related entity.

In this case our Keyless Entity Type is trying to map a collection navigation property, so our assumption is that's the cause.

---

### Expected Result
If is this is an explicit design decision not to support collection navigation properties on Keyless Entity Types, then the messaging should be improved to state what the problem is (with references to the applicable documentation). Errors like this can take quite a while to investigate, and good messaging greatly improves the customer experience. I note that https://github.com/dotnet/efcore/issues/20292 was closed, which looks to be of a similar concern.

More broadly:
- Has the reasoning behind not supporting this use case been documented or discussed anywhere? 
- Is there a plan to support it moving forwards? 
- Is there a suggested workaround?

Many thanks, and looking forward to hearing your thoughts :)

### Classes
```C#
public class vwContainerSource // Keyless Entity Type
{
    public vwContainerSource()
    {
        ShippingLineContainerTariff = new HashSet<ShippingLineContainerTariff>();
    }

    public int CompanyShippingId { get; set; }
    
    [Required]
    [StringLength(50)]
    public string ContainerNumber { get; set; }
    
    public virtual Company CompanyShipping { get; set; }
    
    public virtual ICollection<ShippingLineContainerTariff> ShippingLineContainerTariff { get; set; }
}

public class ShippingLineContainerTariff // Regular entity
{
    public int ShippingLineContainerTariffId { get; set; }
    
    public int CompanyShippingId { get; set; }
    
    [Required]
    [StringLength(50)]
    public string ContainerNumber { get; set; }
    
    public virtual Company CompanyShipping { get; set; }
}
```

### Mapping
```C#
modelBuilder.Entity<vwContainerSource>(entity =>
    entity
        .HasNoKey()
        .ToView("vwContainerSource")
        .HasMany(d => d.ShippingLineContainerTariff)
        .WithOne()
        .HasForeignKey(d => new { d.CompanyShippingId, d.ContainerNumber })
        .HasPrincipalKey(d => new { d.CompanyShippingId, d.ContainerNumber })
    );
```

### Result
```
System.NullReferenceException: 'Object reference not set to an instance of an object.'
   at Microsoft.EntityFrameworkCore.Metadata.Internal.InternalEntityTypeBuilder.SetOrAddForeignKey(ForeignKey foreignKey, InternalEntityTypeBuilder principalEntityTypeBuilder, IReadOnlyList`1 dependentProperties, Key principalKey, String navigationToPrincipalName, Nullable`1 isRequired, Nullable`1 configurationSource)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.InternalEntityTypeBuilder.CreateForeignKey(InternalEntityTypeBuilder principalEntityTypeBuilder, IReadOnlyList`1 dependentProperties, Key principalKey, String navigationToPrincipalName, Nullable`1 isRequired, ConfigurationSource configurationSource)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.InternalEntityTypeBuilder.HasRelationshipInternal(EntityType targetEntityType, Key principalKey, ConfigurationSource configurationSource)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.InternalEntityTypeBuilder.HasRelationship(EntityType principalEntityType, ConfigurationSource configurationSource)
   at Microsoft.EntityFrameworkCore.Metadata.Builders.EntityTypeBuilder`1.HasMany[TRelatedEntity](Expression`1 navigationExpression)
   at Redacted.<>c.<OnModelCreatingViews>b__348_0(EntityTypeBuilder`1 entity) in Redacted\Entities.cs:line 1374
   at Microsoft.EntityFrameworkCore.ModelBuilder.Entity[TEntity](Action`1 buildAction)
   at Redacted.OnModelCreatingViews(ModelBuilder modelBuilder) in Redacted\Entities.cs:line 1373
   at Redacted.OnModelCreating(ModelBuilder modelBuilder) in Redacted\Entities.cs:line 1365
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelCustomizer.Customize(ModelBuilder modelBuilder, DbContext context)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_3(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_Model()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityType()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.CheckState()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityQueryable()
   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.System.Linq.IQueryable.get_Provider()
   at System.Linq.Queryable.Where[TSource](IQueryable`1 source, Expression`1 predicate)
   at Redacted
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
```

### Include provider and version information

EF Core version: 3.1.5 [this also repros in 3.1.12]
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3
Operating system: Windows 10
IDE: Visual Studio 2019 16.5

**Comment by anranruye** [comment]
There is a conflict between `.HasNoKey()` and `.HasPrincipalKey()` which will define an alternate key.

**Comment by optiks** [comment]
> There is a conflict between `.HasNoKey()` and `.HasPrincipalKey()`(which will define an alternate key).

Thanks for your reply, but that's not the problem: it's the call to `.HasMany()` that fails (it doesn't even get to the `.HasPrincipalKey()`).

**Comment by anranruye** [comment]
@optiks 
Yes, as the documentation, keyless entity can only be the independent side in a relationship. Just notice that a child entity can not find its parent if the parent has no key, can't it?

I want to point that, you should use regular entity for `vwContainerSource` class since it actually has key properties. Also, you should change your migration file manully if you use migration, by removing the table and foriegn key creation code.

**Comment by optiks** [comment]
@anangaur Thanks again for your reply. 

> Yes, as the documentation, keyless entity can only be the independent side in a relationship. Just notice that a child entity can not find its parent if the parent has no key, can't it?

As I mentioned in Expected Result, the bug _isn't that this scenario isn't supported_, it's that the result of invoking the unsupported path is a `NullReferenceException` -- a terrible experience which then leads to unnecessary time being spent trying to work out what's going on. I'm sure the EF team is aware of this complaint, as bugs frequently manifest as obscure exceptions.

> I want to point that, you should use regular entity for `vwContainerSource` class since it actually has key properties. Also, you should change your migration file manully if you use migration, by removing the table and foriegn key creation code.

We're already using "regular entities" as a workaround, but it feels like a hack, to me. We want the standard Keyless Entity Type benefits (changes aren't tracked, not discovered by convention, HasNoKey), and I don't really see _why_ having a primary key defined (which isn't even used in the relationship) is necessary?

**Comment by lakeman** [comment]
An sql query that returns both parent and child properties, must be able to step through the flat result set, recognise which rows in the result set map to the same parent object, materialise the child objects and link them to that parent. This is impossible without a key on the parent.

**Comment by optiks** [comment]
> An sql query that returns both parent and child properties, must be able to step through the flat result set, recognise which rows in the result set map to the same parent object, materialise the child objects and link them to that parent. This is impossible without a key on the parent. Sure, a better error would be good though.

Thanks, good reply. I haven't put too much thought into it, but it does still _seem_ like `.HasPrincipalKey(d => new { d.CompanyShippingId, d.ContainerNumber })` could work? Let's use an example.

| vwContainerSource   
| ------------- | 
| { CompanyShippingId: 1, ContainerNumber: "ABCD1234567, "Type": "A" }  | 
| { CompanyShippingId: 1, ContainerNumber: "ABCD1234567, "Type": "B" }  | 
| { CompanyShippingId: 1, ContainerNumber: "AAAA1111111, "Type": "C" }  |

| ShippingLineContainerTariff |
 ------------- |
| { CompanyShippingId: 1, ContainerNumber: "ABCD1234567", "Type": "Detention" }  |
| { CompanyShippingId: 1, ContainerNumber: "ABCD1234567", "Type": "Demurrage" }  |

Steps:
1. Step through result set
2. Materialise parent objects 
3. Materialise child objects - link to any parent that matches on `{ CompanyShippingId, ContainerNumber }` (within _this query scope_ - no general entity tacking). Yes, there could be _multiple_ instances of the parent, but why does that matter? The link on `{ CompanyShippingId, ContainerNumber }` still holds true, right?

Isn't the following output still valid based on the relationships above?
* `{ CompanyShippingId: 1, ContainerNumber: "ABCD1234567, "Type": "A" }` => `[ "Detention", "Demurrage" ]`
* `{ CompanyShippingId: 1, ContainerNumber: "ABCD1234567, "Type": "B" }` => `[ "Detention", "Demurrage" ]` // same entity references as above

Am I missing something?

**Comment by ajcvickers** [comment]
@optiks Nice to talk to you again. :-) First, the `NullReferenceException` is a bug. We should be detecting the situation and throwing a better exception message.

Second, EF Core needs to be able to uniquely identity the parent entity. If such a set of columns exist, then they should be defined as the key for the parent type. It doesn't matter if their is such a key in the database or not; it just needs to provide a unique identifier for each row read.

Once the principal type is no longer keyless, then a relationship can be defined to the child in the normal way. Again, it doesn't matter if that relationship is defined in the database or not.

**Comment by optiks** [comment]
@ajcvickers Thanks - better messaging would be greatly appreciated.

> Second, EF Core needs to be able to uniquely identity the parent entity. If such a set of columns exist, then they should be defined as the key for the parent type. It doesn't matter if their is such a key in the database or not; it just needs to provide a unique identifier for each row read.

I get that, but by switching to regular entities it means taking on additional behaviours that we don't really want (primarily entity tracking). 
1. Did you see my (no doubt extremely naive) suggestion above?
2. Is there a way of opting out of the additional behaviours that get tacked on? Making `vwContainerSource` read-only would probably be good enough. https://github.com/dotnet/efcore/issues/7586

**Comment by ajcvickers** [comment]
@optiks If you don't want the entities to be tracked, then use a no-tracking query. We don't recommend using keyless entity types just so they are not tracked.

**Comment by optiks** [comment]
@ajcvickers Okie - I'll let it go :-).

The reality is that the view result _is not an entity_ (in fact, for us it represents the unification of three different entities) - so it's more of an ideological complaint than a practical one. Pushing no-tracking to the caller is less-than-ideal, but it works.

**Comment by ajcvickers** [comment]
@optiks There is inherently no difference between a CLR typed mapped to a view and a CLR type mapped to a table. Either could be updatable or read-only. Either could have a primary key defined, or not have a primary key defined. And so on. Therefore, we ultimately decided to call them all "entity types", because otherwise people read too much semantics into the naming. For example, "view types" where assumed to be read-only because people usually use read-only views. Likewise, people assumed views could not be mapped to types with keys because views don't usually have keys.

**Comment by smitpatel** [comment]
Error message in current 6.0 branch
System.InvalidOperationException : The keyless entity type 'KeylessCollectionNavigation' cannot be on the principal end of the relationship between 'KeylessCollectionNavigation.Stores' and 'Store'. The principal entity type must have a key.

**Comment by smitpatel** [comment]
Adding regression test to main branch. Leaving issue in 6.0 since the bug is fixed in 6.0 already.
